Added userdata page and more factories

kdmnk opened the pull request:
Added page for changing email and password.

- Might need to send confirmation email (?)
- Tested email changing for identical emails

- Backend for password changing is missing

__________
_(outdated)_

luksan47 commented:
> Backend for password changing is missing

It's a Laravel built-in, working properly.

luksan47 commented:
Looks nice! Once everything is done, this will be a nice feature! Added some modifications to help you out.
(unfortunately cannot add to the release now)

luksan47 commented:
@kdmnk what do you think, after rebasing, are we close to finish this?

kdmnk commented:
I will have a look at it

luksan47 commented on app/Http/Controllers/UserController.php:
```diff
@@ -61,5 +61,43 @@ public function updateEmail(Request $request)
         return redirect()->route('userdata');
     }
-    //TODO updatePassword 
+    public function changePassword(Request $request){
```
Isn't there a Laravel built-in for this?

kdmnk replied:
> Shouln't be there also the personal information that the user filled out when registering?

Yes, I am working on it. I was thinking about showing the phone number, neptun code, and the faculties and workshops. But `{{$user->neptun}}` and `{{$user->phone_number}}` is not working as expected as with  `{{$user->name}}`. Do you have any idea why?

kdmnk replied:
> Isn't there a Laravel built-in for this?

As I looked up on this a little, I think there's only password reset built-in (with email verification etc...)

luksan47 commented:
Don't forget to change to target branch to uran-2.x.

luksan47 commented on app/Http/Controllers/UserController.php:
```diff
@@ -2,10 +2,75 @@
+        $user = Auth::user();
+
+        $validator = Validator::make($request->all(), [
+            'old_password' => 'required|string|same:',
```
I think the root cause of the problem is here: `same:` missing parameter.

luksan47 commented on resources/views/auth/user.blade.php:
```diff
@@ -6,15 +6,110 @@
+                        {{ $message }}
+                    </blockquote>
+                    @enderror
+                </form>
```
nit: extra </form>

luksan47 commented on resources/views/layouts/app.blade.php:
```diff
@@ -42,10 +42,15 @@ function() {
                 });
+                $('.collapsible').collapsible();
+                @if (session('message'))
+                console.log('ok');
```
nit: remove console.log

kdmnk commented:
Please test everything again.

luksan47 commented on app/Http/Controllers/UserController.php:
```diff
@@ -54,11 +54,11 @@ public function updatePassword(Request $request)
             'new_password' => 'required|string|min:8|confirmed|different:old_password',
         ]);
-
+        $validator->validate();
```
Actually, I just forgot to remove this part, sorry. You can keep either that or the one below.

luksan47 commented on app/Http/Controllers/UserController.php:
```diff
@@ -2,10 +2,94 @@
+        $user = Auth::user();
+
+        return view('auth.user', ['user' => $user])
+            ->with('neptun', $user->educationalInformation->neptun)
```
Not all users have educationalInformation, for them it fails. Also, you do not have to pass any of these values, as you have access to Auth::user() in the blade.

luksan47 commented on app/Http/Controllers/UserController.php:
```diff
@@ -2,10 +2,94 @@
+        $validator = Validator::make($request->all(), [
+            'email' => 'required|email|max:225|unique:users',
+        ]);
+        $validator->validate();
```
Same here as L83.

luksan47 commented on app/Http/Controllers/UserController.php:
```diff
@@ -2,10 +2,94 @@
+            'email' => $request->email,
+        ]);
+
+        return redirect()->back()->with('message', 'ok');
```
Is the message used?

kdmnk replied:
Not yet, I want to standardize this in a new issue. For now, I will delete.

luksan47 replied:
Okay, great! You might want to look if there is some existing solution?

luksan47 commented on app/Http/Controllers/UserController.php:
```diff
@@ -2,10 +2,94 @@
+        $validator = Validator::make($request->all(), [
+            'phone_number' => 'required|string|min:16|max:18',
+        ]);
+        $validator->validate();
```
Same as L61.

luksan47 commented on app/Http/Controllers/UserController.php:
```diff
@@ -2,10 +2,94 @@
+            'password' => Hash::make($request->new_password),
+        ]);
+
+        return redirect()->back()->with('message', 'ok');
```
Same, do we use this msg?

luksan47 commented on resources/views/auth/user.blade.php:
```diff
@@ -6,14 +6,155 @@
                 <div class="card-title">@lang('info.user_data')</div>
-                <p>Your user data will appear here</p>
-                <!--TODO-->
+                <table>
```
Does the new UI rely on tables instead of divs? I prefer using divs as they can be configured to look good on all devices.

kdmnk replied:
I tested this on mobiles, it's easier for now, but I agree. (btw there is a responsive-table option in materialize, that swaps columns and rows on smaller sizes, but that's horrible in this case)

luksan47 commented on routes/web.php:
```diff
@@ -31,9 +31,14 @@
     return view('auth.verification');
 })->name('verification');
 
-Route::middleware(['auth', 'log'])->group(function () {
```
I would only move the pw route to an unlogged group.

kdmnk commented:
In the case when the user doesn't have personalInformation (which should not happen), the updatePhone function will fail. What should I do with it?

kdmnk commented:
Updated collegist and tenant seeders also with random data.

luksan47 commented on app/Http/Controllers/UserController.php:
```diff
@@ -2,10 +2,90 @@
+        $user->personalInformation->update([
+            'phone_number' => $request->phone_number,
+        ]);
+        //if user doesn`t have personalInformation TODO
```
How about simply creating instead of updating for those without personalInformation? That would be a simple solution that won't give an error.

luksan47 replied:
Or you could simply hide the fields for these users, might be even better.

kdmnk replied:
> How about simply creating instead of updating for those without personalInformation? That would be a simple solution that won't give an error.

Yes, but we don't have all the information for that. I will try to hide it.

luksan47 replied:
You can add a member function to the user model:
`public function hasPersonalInformation(){
  return $this-> personalInformation === null;
}`
Haven't tested.

kdmnk replied:
Yes, but quite the opposite :)

luksan47 replied:
Yeah, but you can negate it :)

luksan47 commented on database/seeds/UsersTableSeeder.php:
```diff
@@ -55,6 +56,30 @@ private function createCollegist()
         $user->roles()->attach(Role::getId(Role::PRINTER));
         $user->roles()->attach(Role::getId(Role::INTERNET_USER));
         $user->internetAccess->setWifiUsername();
+        PersonalInformation::create([
```
Please create a factory here. If we write a test, we would use it anyway and it's safer to see randomly generated data. These are pretty simple, you can check the others in /factories/.

luksan47 commented on database/seeds/UsersTableSeeder.php:
```diff
@@ -55,6 +56,30 @@ private function createCollegist()
+            'city' => 'Budapest',
+            'street_and_number' => 'Ménesi út 11-13',
+        ]);
+        App\EducationalInformation::create([
```
Same here as above.

kdmnk commented:
Made factories for EducationalInformation and PersonalInformation. Also added random workshops and faculties for random users. (note that a user can have multiple workshops/faculties of the same type this way).
Also added these for predefined test users.

luksan47 commented on database/seeds/UsersTableSeeder.php:
```diff
@@ -18,12 +18,31 @@ public function run()
+            for ($x = 0; $x < rand(1,3); $x++) {
+                $user->faculties()->attach(rand(1,7));
+            }
+            for ($x = 0; $x < rand(1,3); $x++) {
```
You could use the same loop here.

kdmnk replied:
I used different loops to make it more random (this way, the user can have different number of faculties and workshops)

luksan47 commented on database/seeds/UsersTableSeeder.php:
```diff
@@ -18,12 +18,31 @@ public function run()
+                $user->faculties()->attach(rand(1,7));
+            }
+            for ($x = 0; $x < rand(1,3); $x++) {
+                $user->workshops()->attach(rand(1,17));
```
Instead of hardcoded numbers, you could get the length of he lists from the respective models, here and above.

luksan47 commented on database/seeds/UsersTableSeeder.php:
```diff
@@ -38,9 +57,18 @@ private function createAdmin()
+        factory(App\EducationalInformation::class, $user->id % 5)->create(['user_id' => $user->id]);
+        for ($x = 0; $x < rand(1,3); $x++) {
+            $user->faculties()->attach(rand(1,7));
+        }
```
Same here as above.
